File: samplestore/src/main/java/com/stripe/samplestore/StoreUtils.java
package com.stripe.samplestore;

import android.support.annotation.Nullable;

import java.text.DecimalFormat;
import java.util.Currency;

/**
 * Class for utility functions.
 */
public class StoreUtils {

    static String getEmojiByUnicode(int unicode){
        return new String(Character.toChars(unicode));
    }

    static String getPriceString(long price, @Nullable Currency currency) {
        Currency displayCurrency = currency == null
                ? Currency.getInstance("USD")
                : currency;

        int fractionDigits = displayCurrency.getDefaultFractionDigits();
        int totalLength = String.valueOf(price).length();
        StringBuilder builder = new StringBuilder();
        builder.append('\u00A4');

        if (fractionDigits == 0) {
            for (int i = 0; i < totalLength; i++) {
                builder.append('#');
            }
            DecimalFormat noDecimalCurrencyFormat = new DecimalFormat(builder.toString());
            noDecimalCurrencyFormat.setCurrency(displayCurrency);
            return noDecimalCurrencyFormat.format(price);
        }

        int beforeDecimal = totalLength - fractionDigits;
        for (int i = 0; i < beforeDecimal; i++) {
            builder.append('#');
        }
        // So we display "$0.55" instead of "$.55"
        if (totalLength <= fractionDigits) {
            builder.append('0');
        }
        builder.append('.');
        for (int i = 0; i < fractionDigits; i++) {
            builder.append('0');
        }
        double modBreak = Math.pow(10, fractionDigits);
        double decimalPrice = price / modBreak;

        DecimalFormat decimalFormat = new DecimalFormat(builder.toString());
        decimalFormat.setCurrency(displayCurrency);

        return decimalFormat.format(decimalPrice);
    }
}

File: samplestore/src/main/java/com/stripe/samplestore/StoreUtils.kt
package com.stripe.samplestore

import java.text.DecimalFormat
import java.util.Currency
import kotlin.math.pow

/**
 * Class for utility functions.
 */
internal object StoreUtils {

    private const val CURRENCY_SIGN = '\u00A4'

    fun getEmojiByUnicode(unicode: Int): String {
        return String(Character.toChars(unicode))
    }

    fun getPriceString(price: Long, currency: Currency?): String {
        val displayCurrency = currency ?: Currency.getInstance("USD")

        val fractionDigits = displayCurrency.defaultFractionDigits
        val totalLength = price.toString().length
        val builder = StringBuilder().append(CURRENCY_SIGN)

        if (fractionDigits == 0) {
            for (i in 0 until totalLength) {
                builder.append('#')
            }
            val noDecimalCurrencyFormat = DecimalFormat(builder.toString())
            noDecimalCurrencyFormat.currency = displayCurrency
            return noDecimalCurrencyFormat.format(price)
        }

        val beforeDecimal = totalLength - fractionDigits
        for (i in 0 until beforeDecimal) {
            builder.append('#')
        }
        // So we display "$0.55" instead of "$.55"
        if (totalLength <= fractionDigits) {
            builder.append('0')
        }
        builder.append('.')
        for (i in 0 until fractionDigits) {
            builder.append('0')
        }
        val modBreak = 10.0.pow(fractionDigits.toDouble())
        val decimalPrice = price / modBreak

        val decimalFormat = DecimalFormat(builder.toString())
        decimalFormat.currency = displayCurrency

        return decimalFormat.format(decimalPrice)
    }
}
